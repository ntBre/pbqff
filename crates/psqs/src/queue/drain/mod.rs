use core::time;
use std::{
    borrow::BorrowMut,
    collections::{HashMap, HashSet},
    iter::{Enumerate, Fuse, Peekable},
    slice::ChunksMut,
    sync::LazyLock,
    thread,
};

use crate::{
    geom::Geom,
    program::{Job, Procedure, Program, ProgramResult},
    queue::drain::{dump::Dump, resub::ResubOutput},
};

use super::Queue;

/// time the duration of `$body` and store the resulting Duration in `$elapsed`
#[macro_export]
macro_rules! time {
    ($elapsed:ident, $body:block) => {
        let now = std::time::Instant::now();
        $body;
        let $elapsed = now.elapsed();
    };
}

mod dump;
mod resub;
mod timer;

use libc::{timeval, RUSAGE_SELF};
use resub::Resub;
use serde::{Deserialize, Serialize};

static NO_RESUB: LazyLock<bool> =
    LazyLock::new(|| std::env::var("NO_RESUB").is_ok());

pub enum Check {
    Some { check_int: usize, check_dir: String },
    None,
}

pub(crate) trait Drain {
    type Item;

    fn procedure(&self) -> Procedure;

    fn set_result<P: Program>(
        &self,
        dst: &mut [Self::Item],
        job: &mut Job<P>,
        res: ProgramResult,
    );

    /// on success, return the total job time, as returned by `P::read_output`.
    /// on failure, return a vector of failed job indices
    fn drain<P, Q>(
        &self,
        dir: &str,
        queue: &Q,
        mut jobs: Vec<Job<P>>,
        dst: &mut [Self::Item],
        check: Check,
    ) -> Result<f64, Vec<usize>>
    where
        Self: Sync,
        P: Program + Clone + Send + Sync + Serialize + for<'a> Deserialize<'a>,
        Q: Queue<P> + ?Sized + Sync,
        <Self as Drain>::Item: Clone + Serialize,
    {
        // total time for the jobs to run as returned from Program::read_output
        let mut job_time = 0.0;

        let mut cur_jobs = Vec::new();
        let mut slurm_jobs = HashMap::new();
        let mut remaining = jobs.len();

        let job_limit = queue.job_limit();

        let mut out_of_jobs = false;

        let dump = Dump::new(queue.no_del());
        let mut time = timer::Timer::default();

        let mut qstat = HashSet::<String>::new();
        // this is a bit sad, but I need the original jobs for checkpoints and I
        // can't get an immutable reference to them while chunks is holding a
        // mutable reference. also can't use a Cow because the chunks_mut call
        // would clone immediately anyway. I wanted something like Cow because I
        // only need this clone if check_int is nonzero. I also can't figure out
        // how to clone chunks itself when writing the checkpoint. that would
        // really be the ideal solution, but it seems I can only consume the
        // iterator. another option would be to consume the iterator and rebuild
        // it when writing the checkpoints
        let jobs_init = if let Check::Some { .. } = check {
            jobs.clone()
        } else {
            Vec::new()
        };
        let total_jobs = jobs.len();
        // for fast jobs, it may be necessary to stop and clean up even if
        // finished != 0. this is used to signal that case
        let mut cleanup_intervals =
            (0..total_jobs).step_by(job_limit).peekable();

        let mut chunks = jobs
            .chunks_mut(queue.chunk_size())
            .enumerate()
            .fuse()
            .peekable();
        // the index of the last chunk consumed. used for writing remaining jobs
        // to checkpoints. None initially and then Some(chunk_num)
        let mut last_chunk = None;
        let mut to_remove = Vec::new();
        let mut resub = Resub::new(queue, dir, self.procedure());
        let mut failed_jobs = HashSet::new();
        let mut iter = 0;
        const MAX_RETRIES: usize = 5;
        let mut retries = HashMap::new();
        loop {
            let loop_time = std::time::Instant::now();
            if chunks.peek().is_none() {
                out_of_jobs = true;
            }
            if !out_of_jobs {
                let n = self.receive_jobs(
                    &mut chunks,
                    job_limit,
                    &mut cur_jobs,
                    queue,
                    dir,
                    &mut slurm_jobs,
                    &mut time,
                    &mut qstat,
                    &mut last_chunk,
                );
                log::trace!("received {n} chunks of jobs");
            }

            // collect output
            let mut finished = 0;
            to_remove.clear();
            let now = std::time::Instant::now();
            let outfiles: Vec<_> =
                cur_jobs.iter().map(|job| job.program.filename()).collect();
            use rayon::prelude::*;
            let results: Vec<_> =
                outfiles.par_iter().map(|out| P::read_output(out)).collect();
            time.reading += now.elapsed();
            for (i, (job, res)) in cur_jobs.iter_mut().zip(results).enumerate()
            {
                match res {
                    Ok(res) => {
                        let name = job.program.filename();
                        if failed_jobs.remove(&name) {
                            log::info!("removed {name} from failed_jobs");
                        }
                        to_remove.push(i);
                        job_time += res.time;
                        self.set_result(dst, job, res);
                        for f in job.program.associated_files() {
                            dump.send(f);
                        }
                        finished += 1;
                        remaining -= 1;
                        let job_name = job.pbs_file.as_str();
                        let mut count = match slurm_jobs.get_mut(job_name) {
                            Some(n) => *n,
                            None => {
                                eprintln!(
                                    "failed to find {job_name} in slurm_jobs"
                                );
                                1
                            }
                        };
                        count -= 1;
                        if count == 0 {
                            // delete the submit script and output file
                            dump.send(job_name.to_string());
                            dump.send(format!("{job_name}.out"));
                        }
                    }
                    Err(e) => {
                        if e.is_error_in_output() {
                            let filename = job.program.filename();
                            if !failed_jobs.contains(&filename) {
                                log::warn!("job failed with `{e}`");
                                failed_jobs.insert(filename);
                            }
                        } else if !qstat.contains(&job.job_id) {
                            // to avoid temporary file system issues, check a
                            // few times before resubmitting. this should avoid
                            // the case I've been seeing where I end up with 70+
                            // .out_## files in Molpro. the jobs are obviously
                            // running and finishing, but the output files
                            // aren't appearing until after I've already
                            // resubmitted
                            let retry = retries
                                .entry(job.program.filename())
                                .or_insert(MAX_RETRIES);
                            if *retry == 0 {
                                // just overwrite the existing job with
                                // the resubmitted version
                                let time = job.modtime();
                                if time > job.modtime {
                                    // file has been updated since we last
                                    // looked at it, so need to look again
                                    job.modtime = time;
                                } else {
                                    // actual resubmission path
                                    eprintln!(
                                        "resubmitting {} (id={}) for {:?}",
                                        job.program.filename(),
                                        job.job_id,
                                        e
                                    );
                                    if *NO_RESUB {
                                        eprintln!(
                                            "resubmission disabled by \
					 NO_RESUB environment \
					 variable, exiting"
                                        );
                                        std::process::exit(1);
                                    }
                                    failed_jobs.remove(&job.program.filename());
                                    // copy the job into resub and plan to
                                    // remove it from cur_jobs
                                    resub.push(job.clone());
                                    to_remove.push(i);
                                }
                            } else {
                                *retry -= 1;
                            }
                        }
                    }
                }
            }
            // have to remove the highest index first so sort and reverse
            let r = std::time::Instant::now();
            to_remove.sort();
            to_remove.reverse();
            for i in &to_remove {
                cur_jobs.swap_remove(*i);
            }
            time.removing += r.elapsed();
            // submit resubs
            let works = resub.resubmit();
            for ResubOutput {
                jobs,
                slurm_jobs: sj,
                job_id,
                writing_input: wi,
                writing_script: ws,
                submitting: ss,
            } in works
            {
                slurm_jobs.extend(sj);
                time.writing_input += wi;
                time.writing_script += ws;
                time.submitting_script += ss;
                qstat.insert(job_id);
                cur_jobs.extend(jobs);
            }
            log::debug!(
                "finished {} jobs in {:.1} s",
                finished,
                loop_time.elapsed().as_millis() as f64 / 1000.0
            );
            if cur_jobs.len().saturating_sub(failed_jobs.len()) == 0
                && out_of_jobs
            {
                dump.shutdown();
                if !failed_jobs.is_empty() {
                    if let Check::Some { check_dir, .. } = &check {
                        Self::do_checkpoint(
                            &cur_jobs,
                            last_chunk,
                            &jobs_init,
                            queue.chunk_size(),
                            check_dir,
                            dst,
                        );
                    }
                    let failed_indices: Vec<_> =
                        cur_jobs.into_iter().map(|job| job.index).collect();
                    return Err(failed_indices);
                }
                eprintln!("{time}");
                return Ok(job_time);
            }
            if finished == 0 {
                wait(queue, &mut time, iter, remaining);
                qstat = queue.status();
            } else if total_jobs - remaining
                > *cleanup_intervals.peek().unwrap_or(&total_jobs)
            {
                wait(queue, &mut time, iter, remaining);
                cleanup_intervals.next();
            }
            if let Check::Some {
                check_int,
                check_dir,
            } = &check
            {
                if *check_int > 0 && iter % check_int == 0 {
                    Self::do_checkpoint(
                        &cur_jobs,
                        last_chunk,
                        &jobs_init,
                        queue.chunk_size(),
                        check_dir,
                        dst,
                    );
                }
            }
            iter += 1;
        }
    }

    /// load a checkpoint from the `checkpoint` file, storing the energies in
    /// `dst` and returning the list of remaining jobs
    fn load_checkpoint<P>(
        checkpoint: &str,
        dst: &mut [Self::Item],
    ) -> Vec<Job<P>>
    where
        P: Program + Clone + Send + Sync + Serialize + for<'a> Deserialize<'a>,
        Self::Item: Clone + for<'a> Deserialize<'a>,
    {
        let Ok(f) = std::fs::File::open(checkpoint) else {
            panic!("failed to open {checkpoint}");
        };
        let Checkpoint { dst: d, jobs } = serde_json::from_reader(f).unwrap();
        dst.clone_from_slice(&d);
        jobs
    }

    fn write_checkpoint<P>(
        checkpoint: &str,
        dst: Vec<Self::Item>,
        jobs: Vec<Job<P>>,
    ) where
        P: Program + Clone + Send + Sync + Serialize + for<'a> Deserialize<'a>,
        Self::Item: Serialize,
    {
        let c = Checkpoint { dst, jobs };
        eprintln!("writing checkpoint to {checkpoint}");
        let f = std::fs::File::create(checkpoint).unwrap();
        serde_json::to_writer_pretty(f, &c).unwrap();
    }

    fn do_checkpoint<P>(
        cur_jobs: &[Job<P>],
        last_chunk: Option<usize>,
        jobs_init: &[Job<P>],
        chunk_size: usize,
        check_dir: &str,
        dst: &mut [<Self as Drain>::Item],
    ) where
        P: Program + Clone + Send + Sync + Serialize + for<'a> Deserialize<'a>,
        Job<P>: Clone,
        Self::Item: Serialize + Clone,
    {
        let mut cur_jobs = cur_jobs.to_vec();
        // +1 because after the first chunk (chunk_num = 0) is written,
        // we want to slice from the next chunk on
        let cn = match last_chunk {
            Some(n) => n + 1,
            None => 0,
        };
        cur_jobs.extend(
            jobs_init[(cn * chunk_size).min(jobs_init.len())..].to_vec(),
        );
        Self::write_checkpoint(
            &format!("{check_dir}/chk.json"),
            dst.to_vec(),
            cur_jobs,
        );
    }

    /// Returns the number of chunks received
    #[allow(clippy::too_many_arguments)]
    fn receive_jobs<P, Q>(
        &self,
        chunks: &mut Peekable<Fuse<Enumerate<ChunksMut<Job<P>>>>>,
        job_limit: usize,
        cur_jobs: &mut Vec<Job<P>>,
        queue: &Q,
        dir: &str,
        slurm_jobs: &mut HashMap<String, usize>,
        time: &mut timer::Timer,
        qstat: &mut HashSet<String>,
        last_chunk: &mut Option<usize>,
    ) -> usize
    where
        Self: Sync,
        P: Program + Clone + Send + Sync + Serialize + for<'a> Deserialize<'a>,
        Q: Queue<P> + ?Sized + Sync,
        <Self as Drain>::Item: Clone + Serialize,
    {
        use rayon::prelude::*;
        let works: Vec<_> = chunks
            .borrow_mut()
            .take((job_limit - cur_jobs.len()) / queue.chunk_size())
            // NOTE par_bridge does NOT preserve order
            .par_bridge()
            .map(|(chunk_num, jobs)| {
                let now = std::time::Instant::now();
                let (slurm_jobs, wi, ws, ss) =
                    queue.build_chunk(dir, jobs, chunk_num, self.procedure());
                let job_id = jobs[0].job_id.clone();
                let elapsed = now.elapsed();
                log::debug!(
                    "submitted chunk {} after {:.1} s",
                    chunk_num,
                    elapsed.as_millis() as f64 / 1000.0
                );
                (jobs.to_vec(), slurm_jobs, job_id, wi, ws, ss, chunk_num)
            })
            .collect();
        let ret = works.len();
        for (jobs, sj, job_id, wi, ws, ss, cn) in works {
            slurm_jobs.extend(sj);
            time.writing_input += wi;
            time.writing_script += ws;
            time.submitting_script += ss;
            qstat.insert(job_id);
            cur_jobs.extend(jobs);
            // necessary because par_bridge may swap order
            if let Some(n) = *last_chunk {
                *last_chunk = Some(usize::max(n, cn))
            } else {
                *last_chunk = Some(cn);
            }
        }
        ret
    }
}

fn to_secs(time: timeval) -> f64 {
    time.tv_sec as f64 + time.tv_usec as f64 / 1e6
}

/// return the CPU time used by the current process in seconds
fn get_cpu_time() -> f64 {
    unsafe {
        let mut rusage = std::mem::MaybeUninit::uninit();
        let res = libc::getrusage(RUSAGE_SELF, rusage.as_mut_ptr());
        if res != 0 {
            return 0.0;
        }
        let rusage = rusage.assume_init();
        to_secs(rusage.ru_stime) + to_secs(rusage.ru_utime)
    }
}

fn wait<P, Q>(queue: &Q, time: &mut timer::Timer, iter: usize, remaining: usize)
where
    P: Program + Clone + Send + Sync + Serialize + for<'a> Deserialize<'a>,
    Q: Queue<P> + ?Sized + Sync,
{
    let date = chrono::Local::now().format("%Y-%m-%d %H:%M:%S");
    eprintln!(
        "[iter {iter} {date} {:.1} CPU s] {remaining} jobs remaining",
        get_cpu_time()
    );
    let d = time::Duration::from_secs(queue.sleep_int() as u64);
    time.sleeping += d;
    thread::sleep(d);
}

pub(crate) struct Opt;

impl Drain for Opt {
    type Item = Geom;

    fn procedure(&self) -> Procedure {
        Procedure::Opt
    }

    fn set_result<P: Program>(
        &self,
        dst: &mut [Self::Item],
        job: &mut Job<P>,
        res: ProgramResult,
    ) {
        dst[job.index] = Geom::Xyz(res.cart_geom.unwrap());
    }
}

#[derive(Deserialize, Serialize)]
struct Checkpoint<P, T>
where
    P: Program + Clone,
{
    dst: Vec<T>,
    jobs: Vec<Job<P>>,
}

pub(crate) struct Single;

impl Drain for Single {
    type Item = f64;

    fn procedure(&self) -> Procedure {
        Procedure::SinglePt
    }

    fn set_result<P: Program>(
        &self,
        dst: &mut [Self::Item],
        job: &mut Job<P>,
        res: ProgramResult,
    ) {
        dst[job.index] += job.coeff * res.energy;
    }
}

pub(crate) struct Both;

impl Drain for Both {
    type Item = ProgramResult;

    fn procedure(&self) -> Procedure {
        Procedure::Opt
    }

    fn set_result<P: Program>(
        &self,
        dst: &mut [Self::Item],
        job: &mut Job<P>,
        res: ProgramResult,
    ) {
        dst[job.index] = res;
    }
}
