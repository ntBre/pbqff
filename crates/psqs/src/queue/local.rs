use serde::{Deserialize, Serialize};
use std::collections::HashSet;

use crate::program::dftbplus::DFTBPlus;
use crate::program::molpro::Molpro;
use crate::program::{mopac::Mopac, Program};
use crate::queue::Queue;

use super::{SubQueue, Submit};

/// Minimal implementation for testing MOPAC locally
#[derive(Debug)]
pub struct Local {
    pub dir: String,
    pub chunk_size: usize,
    pub template: Option<String>,
}

impl Default for Local {
    fn default() -> Self {
        Self {
            dir: ".".to_string(),
            chunk_size: 128,
            template: None,
        }
    }
}

impl Local {
    pub fn new(
        chunk_size: usize,
        _job_limit: usize,
        _sleep_int: usize,
        dir: &'static str,
        _no_del: bool,
        template: Option<String>,
    ) -> Self {
        Self {
            dir: dir.to_string(),
            chunk_size,
            template,
        }
    }
}

impl Submit<Molpro> for Local {}

impl Queue<Molpro> for Local {
    fn template(&self) -> &Option<String> {
        &self.template
    }

    fn program_cmd(&self, filename: &str) -> String {
        format!("$MOLPRO_CMD {filename}.inp")
    }

    fn default_submit_script(&self) -> String {
        String::new()
    }
}

impl Submit<Mopac> for Local {}

impl Queue<Mopac> for Local {
    fn template(&self) -> &Option<String> {
        &self.template
    }

    fn program_cmd(&self, filename: &str) -> String {
        format!("$MOPAC_CMD {filename}.mop")
    }

    fn default_submit_script(&self) -> String {
        "export MOPAC_CMD=/opt/mopac/mopac
export LD_LIBRARY_PATH=/opt/mopac/\n"
            .into()
    }
}

impl Submit<DFTBPlus> for Local {}

impl Queue<DFTBPlus> for Local {
    fn template(&self) -> &Option<String> {
        &self.template
    }

    fn program_cmd(&self, filename: &str) -> String {
        format!("(cd {filename} && $DFTB_CMD > out)")
    }

    fn default_submit_script(&self) -> String {
        "DFTB_CMD=/opt/dftb+/dftb+\n".into()
    }
}

impl<P: Program + Clone + Serialize + for<'a> Deserialize<'a>> SubQueue<P>
    for Local
{
    fn submit_command(&self) -> &str {
        "bash"
    }

    fn chunk_size(&self) -> usize {
        self.chunk_size
    }

    fn job_limit(&self) -> usize {
        1600
    }

    fn sleep_int(&self) -> usize {
        1
    }

    const SCRIPT_EXT: &'static str = "slurm";

    fn dir(&self) -> &str {
        &self.dir
    }

    fn stat_cmd(&self) -> String {
        todo!()
    }

    fn status(&self) -> HashSet<String> {
        for dir in ["opt", "pts", "freqs"] {
            let Ok(d) = std::fs::read_dir(dir) else {
                log::error!("{dir} not found for status");
                continue;
            };
            for f in d {
                eprintln!("contents of {:?}", f.as_ref().unwrap());
                eprintln!(
                    "{}",
                    std::fs::read_to_string(f.unwrap().path()).unwrap()
                );
                eprintln!("================");
            }
        }
        panic!("no status available for Local queue");
    }

    fn no_del(&self) -> bool {
        false
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use crate::program::cfour::Cfour;

    use super::*;

    fn local() -> Local {
        Local {
            dir: String::new(),
            chunk_size: 0,
            template: None,
        }
    }

    macro_rules! make_tests {
        ($($name:ident, $queue:expr => $p:ty$(,)*)*) => {
            $(
            #[test]
            fn $name() {
                let tmp = tempfile::NamedTempFile::new().unwrap();
                <Local as Queue<$p>>::write_submit_script(
                    $queue,
                    ["opt0.inp", "opt1.inp", "opt2.inp", "opt3.inp"].map(|s| s.into()),
                    tmp.path().to_str().unwrap(),
                );
                let got = std::fs::read_to_string(tmp).unwrap();
                let got: Vec<&str> = got.lines().filter(|l|
                    !l.contains("/tmp")).collect();
                let got = got.join("\n");
                assert_snapshot!(got);
            }
            )*
        }
    }

    make_tests! {
        mopac_local, &local() =>  Mopac,
        molpro_local, &local() =>  Molpro,
        cfour_local, &local() => Cfour,
        dftb_local, &local() => DFTBPlus,
    }
}
