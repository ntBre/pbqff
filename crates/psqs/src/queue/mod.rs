use std::{
    cell::LazyCell,
    collections::{HashMap, HashSet},
    path::Path,
    process::Command,
    str,
    time::Duration,
};

use crate::{
    geom::Geom,
    program::{Procedure, Program, ProgramError},
};
use crate::{
    program::{Job, ProgramResult},
    time,
};

pub mod local;
pub mod pbs;
pub mod slurm;
use drain::*;
use serde::{Deserialize, Serialize};
mod drain;

pub use drain::Check;

#[derive(PartialEq, Eq, Debug)]
pub struct Resubmit {
    pub inp_file: String,
    pub pbs_file: String,
    pub job_id: String,
}

pub trait Submit<P>: SubQueue<P>
where
    P: Program + Clone + Serialize + for<'a> Deserialize<'a>,
{
    /// submit `filename` to the queue and return the jobid
    fn submit(&self, filename: &str) -> String {
        loop {
            match Command::new(self.submit_command()).arg(filename).output() {
                Ok(s) => {
                    if s.status.success() {
                        let raw = str::from_utf8(&s.stdout)
                            .unwrap()
                            .trim()
                            .to_string();
                        return raw
                            .split_whitespace()
                            .last()
                            .unwrap_or("")
                            .to_string();
                    }
                    eprintln!(
                        "failed to submit {filename} with `{}`",
                        String::from_utf8_lossy(&s.stderr)
                    );
                    std::thread::sleep(Duration::from_secs(1));
                }
                Err(e) => panic!("{e:?}"),
            };
        }
    }
}

/// a trait for all of the program-independent parts of a [Queue]
pub trait SubQueue<P>
where
    P: Program + Clone + Serialize + for<'a> Deserialize<'a>,
{
    /// the extension to append to submit scripts for this type of Queue
    const SCRIPT_EXT: &'static str;

    fn dir(&self) -> &str;

    fn submit_command(&self) -> &str;

    fn chunk_size(&self) -> usize;

    fn job_limit(&self) -> usize;

    fn sleep_int(&self) -> usize;

    /// the command to check the status of jobs in the queue
    fn stat_cmd(&self) -> String;

    /// return a HashSet of jobs found in the queue based on the output of
    /// `stat_cmd`
    fn status(&self) -> HashSet<String>;

    /// return `true` if all output files should be preserved
    fn no_del(&self) -> bool;
}

pub trait Queue<P>: SubQueue<P> + Submit<P>
where
    P: Program + Clone + Send + Sync + Serialize + for<'a> Deserialize<'a>,
{
    fn default_submit_script(&self) -> String;

    fn template(&self) -> &Option<String>;

    fn program_cmd(&self, filename: &str) -> String;

    fn write_submit_script(
        &self,
        infiles: impl IntoIterator<Item = String>,
        filename: &str,
    ) {
        use std::fmt::Write;
        let path = Path::new(filename);
        let basename = path.file_name().unwrap();
        let mut body = self
            .template()
            .clone()
            .unwrap_or_else(|| <Self as Queue<P>>::default_submit_script(self))
            .replace("{{.basename}}", basename.to_str().unwrap())
            .replace("{{.filename}}", filename);
        for f in infiles {
            writeln!(body, "{}", self.program_cmd(&f)).unwrap();
        }
        if std::fs::write(filename, body).is_err() {
            panic!("write_submit_script: failed to create {filename}");
        };
    }

    /// take a name of a Program input file with the extension attached, replace
    /// the extension (ext) with _redo.ext and write _redo.SCRIPT_EXT, then
    /// submit the redo script
    fn resubmit(&self, filename: &str) -> Resubmit {
        let path = Path::new(filename);
        let dir = path.parent().unwrap().to_str().unwrap();
        let base = path.file_stem().unwrap().to_str().unwrap();
        {
            let ext = path.extension().unwrap().to_str().unwrap();
            let inp_file = format!("{dir}/{base}_redo.{ext}");
            if let Err(e) = std::fs::copy(filename, &inp_file) {
                panic!("failed to copy {filename} to {inp_file} with `{e}`")
            }
        }
        // nothing but the copy needs the name with extension
        let inp_name = format!("{dir}/{base}_redo");
        let pbs_file = format!("{}/{}_redo.{}", dir, base, Self::SCRIPT_EXT);
        self.write_submit_script([inp_name.clone()], &pbs_file);
        let job_id = self.submit(&pbs_file);
        Resubmit {
            inp_file: inp_name,
            pbs_file,
            job_id,
        }
    }

    /// Build a chunk of jobs by writing the Program input file and the
    /// corresponding submission script and then submitting the script. returns
    /// the total durations spent writing input files, writing the submit
    /// script, and submitting the script
    fn build_chunk(
        &self,
        dir: &str,
        jobs: &mut [Job<P>],
        chunk_num: usize,
        proc: Procedure,
    ) -> (HashMap<String, usize>, Duration, Duration, Duration) {
        self.build_chunk_inner(dir, "main", chunk_num, jobs, proc)
    }

    fn build_chunk_inner(
        &self,
        dir: &str,
        base: &str,
        chunk_num: usize,
        jobs: &mut [Job<P>],
        proc: Procedure,
    ) -> (HashMap<String, usize>, Duration, Duration, Duration) {
        let mut input = Duration::default();
        let mut script = Duration::default();
        let mut submit = Duration::default();
        let queue_file =
            format!("{}/{base}{}.{}", dir, chunk_num, Self::SCRIPT_EXT);
        let jl = jobs.len();
        let mut slurm_jobs = HashMap::new();
        let filenames = jobs.iter_mut().map(|job| {
            time!(e, {
                job.program.write_input(proc);
            });
            input += e;
            job.pbs_file = queue_file.to_string();
            job.program.filename()
        });
        slurm_jobs.insert(queue_file.clone(), jl);
        time!(e, {
            self.write_submit_script(filenames, &queue_file);
        });
        script += e;
        // run jobs
        let job_id;
        time!(e, {
            job_id = self.submit(&queue_file);
        });
        submit += e;
        for job in jobs {
            job.job_id = job_id.clone();
        }
        (slurm_jobs, input, script, submit)
    }

    fn drain_err_case(
        &self,
        e: ProgramError,
        qstat: &mut HashSet<String>,
        slurm_jobs: &mut HashMap<String, usize>,
        job: &mut Job<P>,
    ) {
        let no_resub = LazyCell::new(|| std::env::var("SEMP_RESUB").is_ok());
        // just overwrite the existing job with the resubmitted
        // version
        if !qstat.contains(&job.job_id) {
            let time = job.modtime();
            if time > job.modtime {
                // file has been updated since we last looked at it, so need to
                // look again
                job.modtime = time;
                return;
            }
            eprintln!(
                "resubmitting {} (id={}) for {:?}",
                job.program.filename(),
                job.job_id,
                e
            );
            if *no_resub {
                eprintln!("resubmission disabled by SEMP_RESUB environment variable, exiting");
                std::process::exit(1);
            }
            let resub = format!(
                "{}.{}",
                job.program.filename(),
                job.program.extension()
            );
            let Resubmit {
                inp_file,
                pbs_file,
                job_id,
            } = self.resubmit(&resub);
            job.program.set_filename(&inp_file);
            job.pbs_file = pbs_file.clone();
            slurm_jobs.insert(pbs_file, 1);
            qstat.insert(job_id.clone());
            job.job_id = job_id;
        }
    }

    /// optimize is a copy of drain for optimizing jobs
    fn optimize(
        &self,
        dir: &str,
        jobs: Vec<Job<P>>,
        dst: &mut [Geom],
    ) -> Result<f64, Vec<usize>>
    where
        Self: Sync,
    {
        Opt.drain(dir, self, jobs, dst, Check::None)
    }

    /// resume draining from the checkpoint file in `checkpoint`
    fn resume(
        &self,
        dir: &str,
        checkpoint: &str,
        dst: &mut [f64],
        check: Check,
    ) -> Result<f64, Vec<usize>>
    where
        Self: Sync,
    {
        let jobs = Single::load_checkpoint(checkpoint, dst);
        eprintln!(
            "resuming from checkpoint in '{checkpoint}' with {} jobs remaining",
            jobs.len()
        );
        self.drain(dir, jobs, dst, check)
    }

    /// run the single-point energy calculations in `jobs`, storing the results
    /// in `dst`. if `check_int` > 0, write checkpoint files at that interval
    fn drain(
        &self,
        dir: &str,
        jobs: Vec<Job<P>>,
        dst: &mut [f64],
        check: Check,
    ) -> Result<f64, Vec<usize>>
    where
        Self: Sync,
    {
        Single.drain(dir, self, jobs, dst, check)
    }

    fn energize(
        &self,
        dir: &str,
        jobs: Vec<Job<P>>,
        dst: &mut [ProgramResult],
    ) -> Result<f64, Vec<usize>>
    where
        Self: Sync,
    {
        Both.drain(dir, self, jobs, dst, Check::None)
    }
}
