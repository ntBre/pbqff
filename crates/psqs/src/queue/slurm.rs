use std::collections::HashSet;

use serde::{Deserialize, Serialize};

use crate::program::mopac::Mopac;
use crate::program::Program;
use crate::program::{dftbplus::DFTBPlus, molpro::Molpro};
use crate::queue::Queue;

use super::{SubQueue, Submit};

/// Slurm is a type for holding the information for submitting a slurm job.
/// `filename` is the name of the Slurm submission script
#[derive(Debug)]
pub struct Slurm {
    chunk_size: usize,
    job_limit: usize,
    sleep_int: usize,
    dir: &'static str,
    no_del: bool,
    pub(crate) template: Option<String>,
}

impl Slurm {
    pub fn new(
        chunk_size: usize,
        job_limit: usize,
        sleep_int: usize,
        dir: &'static str,
        no_del: bool,
        template: Option<String>,
    ) -> Self {
        Self {
            chunk_size,
            job_limit,
            sleep_int,
            dir,
            no_del,
            template,
        }
    }
}

impl<P: Program + Clone + Serialize + for<'a> Deserialize<'a>> Submit<P>
    for Slurm
{
}

impl Queue<Molpro> for Slurm {
    fn template(&self) -> &Option<String> {
        &self.template
    }

    fn program_cmd(&self, filename: &str) -> String {
        format!("$MOLPRO_CMD {filename}.inp")
    }

    fn default_submit_script(&self) -> String {
        include_str!("../../templates/slurm/molpro").to_owned()
    }
}

impl Queue<Mopac> for Slurm {
    fn template(&self) -> &Option<String> {
        &self.template
    }

    fn program_cmd(&self, filename: &str) -> String {
        format!("$MOPAC_CMD {filename}.mop")
    }

    fn default_submit_script(&self) -> String {
        include_str!("../../templates/slurm/mopac").to_owned()
    }
}

impl Queue<DFTBPlus> for Slurm {
    fn template(&self) -> &Option<String> {
        &self.template
    }

    fn program_cmd(&self, filename: &str) -> String {
        format!("(cd {filename} && $DFTB_CMD > out)")
    }

    fn default_submit_script(&self) -> String {
        String::new()
    }
}

impl<P> SubQueue<P> for Slurm
where
    P: Program + Clone + Serialize + for<'a> Deserialize<'a>,
{
    fn submit_command(&self) -> &str {
        "sbatch"
    }

    fn chunk_size(&self) -> usize {
        self.chunk_size
    }

    fn job_limit(&self) -> usize {
        self.job_limit
    }

    fn sleep_int(&self) -> usize {
        self.sleep_int
    }

    const SCRIPT_EXT: &'static str = "slurm";

    fn dir(&self) -> &str {
        self.dir
    }

    /// run `squeue -u $USER`. form of the output is:
    ///
    ///    JOBID PARTITION   NAME     USER ST        TIME  NODES NODELIST(REASON)
    /// 30627992   compute  c3oh-   mdavis  R 46-17:12:23      1 node2
    fn stat_cmd(&self) -> String {
        let user = std::env::var("USER").expect("couldn't find $USER env var");
        let status = match std::process::Command::new("squeue")
            .args(["-u", &user])
            .output()
        {
            Ok(status) => status,
            Err(e) => panic!("failed to run squeue with {e}"),
        };
        String::from_utf8(status.stdout)
            .expect("failed to convert squeue output to String")
    }

    fn status(&self) -> HashSet<String> {
        let mut ret = HashSet::new();
        // wut?
        let lines = <Slurm as SubQueue<P>>::stat_cmd(self);
        let lines = lines.lines();
        for line in lines {
            if !line.contains("JOBID") {
                let fields: Vec<_> = line.split_whitespace().collect();
                assert!(fields.len() == 8);
                // exclude completing jobs to combat stuck completing bug
                if fields[4] != "CG" {
                    ret.insert(fields[0].to_string());
                }
            }
        }
        ret
    }

    fn no_del(&self) -> bool {
        self.no_del
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use crate::program::cfour::Cfour;

    use super::*;

    fn slurm() -> Slurm {
        Slurm {
            chunk_size: 1,
            job_limit: 1,
            sleep_int: 1,
            dir: "/tmp",
            no_del: false,
            template: None,
        }
    }

    macro_rules! make_tests {
        ($($name:ident, $queue:expr => $p:ty$(,)*)*) => {
            $(
            #[test]
            fn $name() {
                let tmp = tempfile::NamedTempFile::new().unwrap();
                <Slurm as Queue<$p>>::write_submit_script(
                    $queue,
                    ["opt0.inp", "opt1.inp", "opt2.inp", "opt3.inp"].map(|s| s.into()),
                    tmp.path().to_str().unwrap(),
                );
                let got = std::fs::read_to_string(tmp).unwrap();
                let got: Vec<&str> = got.lines().filter(|l|
                    !(l.starts_with("#SBATCH --job-name")
                        || l.starts_with("#SBATCH -o"))).collect();
                let got = got.join("\n");
                assert_snapshot!(got);
            }
            )*
        }
    }

    make_tests! {
        mopac_slurm, &slurm() =>  Mopac,
        molpro_slurm, &slurm() => Molpro,
        cfour_slurm, &slurm() => Cfour,
        dftb_slurm, &slurm() => DFTBPlus,
    }
}
