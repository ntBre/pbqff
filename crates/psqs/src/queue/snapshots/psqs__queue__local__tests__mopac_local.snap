---
source: src/queue/local.rs
expression: got
---
export MOPAC_CMD=/opt/mopac/mopac
export LD_LIBRARY_PATH=/opt/mopac/
$MOPAC_CMD opt0.inp.mop
$MOPAC_CMD opt1.inp.mop
$MOPAC_CMD opt2.inp.mop
$MOPAC_CMD opt3.inp.mop
